File: 1352.product-of-the-last-k-numbers.py
#
# @lc app=leetcode id=1352 lang=python3
#
# [1352] Product of the Last K Numbers
#

# @lc code=start
class ProductOfNumbers:

    def __init__(self):
        self.prefix_product = [1]

    def add(self, num: int) -> None:
        if num == 0:
            # If a 0 is added, all prefix products before become invalid
            self.prefix_product = [1]
        else:
            self.prefix_product.append(self.prefix_product[-1] * num)

    def getProduct(self, k: int) -> int:
        if k >= len(self.prefix_product):
            # k == n also returns 0 because there is a leading 1 in the array
            return 0
        # n = len(self.prefix_product)
        # return self.prefix_product[n - 1] // self.prefix_product[n - 1 - k]
        return self.prefix_product[-1] // self.prefix_product[-k - 1]


# Your ProductOfNumbers object will be instantiated and called as such:
# obj = ProductOfNumbers()
# obj.add(num)
# param_2 = obj.getProduct(k)
# @lc code=end

File: 1352.product-of-the-last-k-numbers.cpp
/*
 * @lc app=leetcode id=1352 lang=cpp
 *
 * [1352] Product of the Last K Numbers
 */
#include <bits/stdc++.h>

using std::vector;

// @lc code=start
class ProductOfNumbers {
    vector<int> prefix_product;

public:
    ProductOfNumbers() {
        prefix_product = {1};
    }
    
    // T: O(1)
    void add(int num) {
        if (num == 0) {
            // If a 0 is added, all prefix products before become invalid
            prefix_product = {1};
        } else {
            prefix_product.push_back(prefix_product.back() * num);
        }
    }
    
    // T: O(1)
    int getProduct(int k) {
        const int n = prefix_product.size();
        if (k >= n) {
            // Either k is too large, or there was a 0 added before.
            // k == n also returns 0 because there's a leading 1 in the array
            return 0;
        }
        // IMPORTANT
        return prefix_product[n - 1] / prefix_product[n - k - 1];
    }
};

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers* obj = new ProductOfNumbers();
 * obj->add(num);
 * int param_2 = obj->getProduct(k);
 */
// @lc code=end

